runtime/src/extensions.rs

```diff
@@ -87,77 +89,95 @@ where
+	// from contract layer to substrate layer.
+	// reference: https://github.com/paritytech/ink-examples/blob/b8d2caa52cf4691e0ddd7c919e4462311deb5ad0/psp22-extension/runtime/psp22-extension-example.rs#L236
+	let overhead = contract_host_weight.debug_message;
 
-	// weight for dispatching the call
-	let dispatch_weight = call.get_dispatch_info().weight;
+	let charged_weight = env.charge_weight(base_weight.saturating_add(overhead))?;
+	log::debug!(target:LOG_TARGET, "{} charged weight: {:?}", LOG_PREFIX, charged_weight);
 
-	// charge weight for the cost of the deserialization and the dispatch
-	let _ = env.charge_weight(base_weight.saturating_add(dispatch_weight))?;
+	// read the input as RuntimeCall
+	let call: <T as SysConfig>::RuntimeCall = env.read_as_unbounded(len)?;
 
-	log::debug!(target:LOG_TARGET, " dispatch inputted RuntimeCall: {:?}", call);
+	let charged_dispatch_weight = env.charge_weight(call.get_dispatch_info().weight)?;
+
+	log::debug!(target:LOG_TARGET, "{} inputted RuntimeCall: {:?}", LOG_PREFIX, call);
 
 	let sender = env.ext().caller();
 	let origin: T::RuntimeOrigin = RawOrigin::Signed(sender.account_id()?.clone()).into();
 
-	// TODO: uncomment once charged_weight is fixed
-	// let actual_weight = call.get_dispatch_info().weight;
-	// env.adjust_weight(charged_weight, actual_weight);
 	let result = call.dispatch(origin);
 	match result {
 		Ok(info) => {
-			log::debug!(target:LOG_TARGET, "dispatch success, actual weight: {:?}", info.actual_weight);
+			log::debug!(target:LOG_TARGET, "{} success, actual weight: {:?}", LOG_PREFIX, info.actual_weight);
+
+			// refund weight if the actual weight is less than the charged weight
+			if let Some(actual_weight) = info.actual_weight {
+				env.adjust_weight(charged_dispatch_weight, actual_weight);
+			}
 		},
 		Err(err) => {
-			log::debug!(target:LOG_TARGET, "dispatch failed: error: {:?}", err.error);
+			log::debug!(target:LOG_TARGET, "{} failed: error: {:?}", LOG_PREFIX, err.error);
 			return Err(err.error);
 		},
 	}
 	Ok(())
 }
 
-pub(crate) fn read_state<T, E>(env: Environment<E, InitState>) -> Result<(), DispatchError>
+fn read_state<T, E>(env: Environment<E, InitState>) -> Result<(), DispatchError>
 where
-	T: pallet_contracts::Config + frame_system::Config,
+	T: pallet_contracts::Config + cumulus_pallet_parachain_system::Config + frame_system::Config,
 	E: Ext<T = T>,
 {
-	const LOG_TARGET: &str = "pop-api::extension::read_state";
+	const LOG_PREFIX: &str = " read_state |";
 
 	let mut env = env.buf_in_buf_out();
 
-	// TODO: Substitute len u32 with pop_api::src::impls::pop_network::StringLimit.
-	// Move StringLimit to pop_api_primitives first.
-	let len: u32 = env.in_len();
-	let key: ParachainSystemKeys = env.read_as_unbounded(len)?;
+	// To be conservative, we charge the weight for reading the input bytes.
+	// However, charge weight is not strictly necessary here as reading a fixed value's weight is included in the contract call.
```

Comment:
Not sure this is entirely accurate. `read_as` does indeed use `read_sandbox_memory_as`, but the description of `read_as` states more about the weight of a chain extension, where the function is located, whereas `read_sandbox_memory_as` is a general purpose host function. I think it depends on the context, so would still be cautious for now.

Reply:
What is meant with a fixed value's weight?